### PR TITLE
feat(symbolicator): Add the ability to manipulate the automatic killswitch for the LPQ

### PR DESCRIPTION
**Pull request by relaxolotl**
Builds off of https://github.com/getsentry/sentry/pull/28565's changes. Broken out of https://github.com/getsentry/sentry/pull/28714 to reduce the size of that PR's diff.

This adds in methods to `RealtimeMetricsStore` to fetch, add, and remove projects from the automated kill switch (automated option?) stored on redis which specifies which projects should be sent to symbolicator's low priority queue. 

This does not include code to actually use these methods. Actual usages of the code in this PR will be included in #28714.

Open questions have been left as TODOs on the code, and feedback on those would be much appreciated. I also haven't done anything explicit to ensure any sort of locking on this value, which may be a problem. I'm currently unsure if this is something that's already taken care of by Redis, or if additional work needs to be done to ensure that this value is being safely manipulated in redis.

### CODE REVIEW

File: src/sentry/processing/realtime_metrics/base.py [hunk 1]
```diff
@@ -36,3 +40,78 @@ def increment_project_event_counter(self, project_id: int, timestamp: float) ->
             pipeline.set(key, 0, nx=True, px=self._counter_ttl)
             pipeline.incr(key)
             pipeline.execute()
+
+    # TODO: do these killswitch helpers belong here, or in a different class? LowPriorityQueueStore?
+    # This probably needs locking
+    def get_lpq_projects(self) -> Set[int]:
+        """
+        Fetches the list of projects that are currently using the low priority queue.
```
Comment (flub) [inline src/sentry/processing/realtime_metrics/base.py:48]: nit: do you mind putting these on the same line?  i think the auto-formatters should do this... but whatever.  i'll never mention it again if you ignore this 😄

File: src/sentry/processing/realtime_metrics/base.py [hunk 2]
```diff
@@ -36,3 +40,78 @@ def increment_project_event_counter(self, project_id: int, timestamp: float) ->
             pipeline.set(key, 0, nx=True, px=self._counter_ttl)
             pipeline.incr(key)
             pipeline.execute()
+
+    # TODO: do these killswitch helpers belong here, or in a different class? LowPriorityQueueStore?
+    # This probably needs locking
+    def get_lpq_projects(self) -> Set[int]:
```
Comment (flub) [inline src/sentry/processing/realtime_metrics/base.py:46]: Do we use this?  And will we need to use it?  I was more expecting that we'd need a `is_project_lqp(self, project_id: int) -> bool`. That's O(1), this is O(n). plus you get rid of some annoying typing issues of this `_to_int()` function.

File: src/sentry/processing/realtime_metrics/base.py [hunk 3]
```diff
@@ -36,3 +40,78 @@ def increment_project_event_counter(self, project_id: int, timestamp: float) ->
             pipeline.set(key, 0, nx=True, px=self._counter_ttl)
             pipeline.incr(key)
             pipeline.execute()
+
+    # TODO: do these killswitch helpers belong here, or in a different class? LowPriorityQueueStore?
```
Comment (flub) [inline src/sentry/processing/realtime_metrics/base.py:44]: We can start putting them here i guess.

File: src/sentry/processing/realtime_metrics/base.py [hunk 4]
```diff
@@ -36,3 +40,78 @@ def increment_project_event_counter(self, project_id: int, timestamp: float) ->
...
+        """
+
+        added = self.inner.sadd(LPQ_MEMBERS_KEY, project_id)
+
+        # Looks like this might already be in the LPQ, or redis failed to add it. This is only a
+        # problem if redis failed to add it.
+        if added == 0 and not self.inner.sismember(LPQ_MEMBERS_KEY, project_id):
+            raise LowPriorityQueueMembershipError(
+                f"Failed to move project ID {project_id} to the low priority queue"
+            )
```
Comment (flub) [inline src/sentry/processing/realtime_metrics/base.py:77]: IIUC this can only happen when the key is not a set.  That basically mean it is a programming error, so maybe we should do a `sentry_sdk.capture_message()` or so in that case directly saying that the key does not appear to be a set if it wasn't added?  That saves having to create a custom error type 😄

File: src/sentry/processing/realtime_metrics/base.py [hunk 5]
```diff
@@ -36,3 +40,78 @@ def increment_project_event_counter(self, project_id: int, timestamp: float) ->
...
+
+    def remove_projects_from_lpq(self, project_ids: Set[int]) -> Set[int]:
+        """
+        Removes projects from the low priority queue.
+
+        This restores all specified projects back to the regular queue, unless they have been
+        manually forced into the low priority queue via the store.symbolicate-event-lpq-always kill
+        switch.
+
+
```
Comment (flub) [inline src/sentry/processing/realtime_metrics/base.py:87]: nit: too many blank lines

File: src/sentry/processing/realtime_metrics/base.py [hunk 6]
```diff
@@ -36,3 +40,78 @@ def increment_project_event_counter(self, project_id: int, timestamp: float) ->
...
+        Returns all projects that have been successfully removed from the low priority queue.
+        """
+        if len(project_ids) == 0:
+            return set()
```
Comment (flub) [inline src/sentry/processing/realtime_metrics/base.py:91]: does this special case actually matter?

File: src/sentry/processing/realtime_metrics/base.py [hunk 7]
```diff
@@ -36,3 +40,78 @@ def increment_project_event_counter(self, project_id: int, timestamp: float) ->
...
+        removed = self.inner.srem(LPQ_MEMBERS_KEY, *project_ids)
+
+        if removed == len(project_ids):
+            return project_ids
+
+        # Looks like only a subset of the project IDs were removed from the LPQ list.
+        in_lpq = self.inner.smembers(LPQ_MEMBERS_KEY)
+        was_removed = project_ids.intersection(in_lpq)
+
+        return was_removed
```
Comment (flub) [inline src/sentry/processing/realtime_metrics/base.py:102]: This is a lot of effort to construct this return value.  Does it matter?  Why not return `None`?

File: src/sentry/processing/realtime_metrics/base.py (hunk 1 above)
Comment (relaxolotl) [reply src/sentry/processing/realtime_metrics/base.py:48]: i need to talk to somebody about setting up vscode to format python properly 🙃 i appreciate you calling these things out for me though, it's not that i'm deliberately ignoring or breaking these conventions so much as it is that i'm unaware of them and accidentally doing incorrect things. i'll admit it's a bit of a terrible excuse..

File: src/sentry/processing/realtime_metrics/base.py (hunk 4 above)
Comment (relaxolotl) [reply src/sentry/processing/realtime_metrics/base.py:77]: so i could be reading the docs incorrectly: my interpretation of the (very sparse) docs is that an error will be thrown by `sadd` if the value at the specified key isn't a set, which means that that'll cascade upwards and hopefully just axe the task. could that be incorrect, and `redis-py`'s way of indicating this problem is to return 0?

File: src/sentry/processing/realtime_metrics/base.py (hunk 6 above)
Comment (relaxolotl) [reply src/sentry/processing/realtime_metrics/base.py:91]: the code below can definitely handle this - it's mostly a preoptimization so we skip talking to redis entirely if for some reason no project IDs are fed in. i'm assuming it's relatively costly to perform that round trip but i could be wrong and this specific check isn't necessary.

File: src/sentry/processing/realtime_metrics/base.py (hunk 7 above)
Comment (relaxolotl) [reply src/sentry/processing/realtime_metrics/base.py:102]: i think it's possible to return `None` if the only way `removed != len(project_ids)` is because some of the project IDs weren't in the LPQ to begin with.

i'm mostly trying to catch the scenario where some projects remain in the LPQ even though they were supposed to be removed, but this could be based off of an incorrect understanding of how redis works. if redis doesn't perform partial updates then this can simply return `None`.

Comment (relaxolotl) [reply src/sentry/processing/realtime_metrics/base.py:102]: update: it looks like it should be safe to assume that the full removal was successful if `self.inner.srem()` successfully executes. the only occasion where `removed < len(project_ids)` is if some of the elements in `project_ids` weren't in the set to begin with.

File: src/sentry/processing/realtime_metrics/base.py (hunk 4 above)
Comment (relaxolotl) [reply src/sentry/processing/realtime_metrics/base.py:77]: a quick parse through the `redis` and `redis-py` source suggests that an exception will get thrown if the value's type isn't a set in an `sadd`, so i'll simplify this code in light of that 👍

File: src/sentry/processing/realtime_metrics/base.py (hunk 1 above)
Comment (relaxolotl) [reply src/sentry/processing/realtime_metrics/base.py:30]: so i did a quick search for `"""` and it looks like the standard in this repository is overwhelmingly triple quotes on their own line instead of having it share the line with the first sentence.

given that, i think i'll push back on this a little and say that these follow this implicit convention in the codebase.

**Review by untitaker — COMMENTED** [review]
since this was assigned to owners-ingest: @loewenheim I think this is too much to review, the diff includes the whole metrics store.

File: src/sentry/processing/realtime_metrics/base.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+        """
+        Moves a project to the low priority queue.
+
+        This forces all symbolication events triggered by the specified project to be redirected to
+        the low priority queue, unless the project is manually excluded from the low priority queue
+        via the `store.symbolicate-event-lpq-never` kill switch.
+        """
+        pass
+
+    def remove_projects_from_lpq(self, project_ids: Set[int]) -> None:
```
[the remark this answers is not in the document]
Comment (relaxolotl) [reply src/sentry/processing/realtime_metrics/base.py:45]: good catch; this is unfortunately isn't symmetric between the two for a few reasons:
- the task responsible for adding projects to the LPQ operates on a per-project basis, and therefore only knows if one project belongs in the queue or not
- the tasks adding projects to the LPQ are planned to be async, meaning that we'd have to block on all of them if we wanted to add a list
- the task to remove projects _does_ know all of the projects that need to be removed

which leads to an add operation that takes only a single int, and a remove operation that removes multiple ints.

[the remark this answers is not in the document]
Comment (relaxolotl) [reply src/sentry/processing/realtime_metrics/base.py:56]: guess i'll have to eat my own words 🙃 

i've added a `remove_project_from_lpq(self, project_id)` in the PR that follows this here: https://github.com/getsentry/sentry/pull/28714/commits/27386f981232372fbb20c83e970f6b951b733f9b

i can move all changes made to the functions introduced here out of that PR into this one if we want, but there are some signature changes that are related to some logging behaviour that i reckon will likely be contentious.

**Review by flub — APPROVED** [review]

File: src/sentry/processing/realtime_metrics/base.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (loewenheim) [reply src/sentry/processing/realtime_metrics/base.py:56]: I think it should be fine this way.

File: src/sentry/processing/realtime_metrics/redis.py [hunk 9]
```diff
@@ -80,3 +84,59 @@ def increment_project_duration_counter(
...
+        Returns a list of project IDs.
+        """
+        return set(
+            filter(
+                None,
+                {
+                    _to_int(project_id_raw)
+                    for project_id_raw in self.cluster.smembers(LPQ_MEMBERS_KEY)
+                },
+            )
```
[the remark this answers is not in the document]
Comment (loewenheim) [reply src/sentry/processing/realtime_metrics/redis.py:101]: I agree with Floris that this function shouldn't filter anything; project ids that aren't (string representation of) ints are indicative of a bug somewhere and shouldn't be silently skipped.

**Review by loewenheim — CHANGES_REQUESTED** [review]

File: src/sentry/processing/realtime_metrics/redis.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (relaxolotl) [reply src/sentry/processing/realtime_metrics/redis.py:101]: the drawback of me breaking this PR up into two is that yep, @flub brought this up a few times in https://github.com/getsentry/sentry/pull/28714#discussion_r713849046.

i made changes there to remove calls to `_to_int` in that PR, but it meant that this still remained here. i'll backport some of those changes into this PR.

File: src/sentry/processing/realtime_metrics/base.py (hunk 2 above)
Comment (relaxolotl) [reply src/sentry/processing/realtime_metrics/base.py:38]: conversation was continued and resolved in https://github.com/getsentry/sentry/pull/28714#discussion_r713891555 as that PR uses this helper.

tl;dr it's the simplest solution right now to get the info we need, but it could definitely be improved if this ends up being a source of performance problems.

File: src/sentry/processing/realtime_metrics/redis.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (relaxolotl) [reply src/sentry/processing/realtime_metrics/redis.py]: ported these changes over and made a comment in the parent PR that might be of interest: https://github.com/getsentry/sentry/pull/28714#discussion_r718850558

tl;dr we can let the automated bits of this system fail instead of trying too hard to keep it running since we have the manual kill switches as a fallback if that happens

@loewenheim could i bother you to take a look at the ported changes?

File: src/sentry/processing/realtime_metrics/redis.py [hunk 10]
```diff
@@ -80,3 +84,59 @@ def increment_project_duration_counter(
...
+        )
+
+    def add_project_to_lpq(self, project_id: int) -> None:
+        """
+        Moves a project to the low priority queue.
+
+        This forces all symbolication events triggered by the specified project to be redirected to
+        the low priority queue, unless the project is manually excluded from the low priority queue
+        via the `store.symbolicate-event-lpq-never` kill switch.
```
[the remark this answers is not in the document]
Comment (relaxolotl) [reply src/sentry/processing/realtime_metrics/redis.py]: i've removed all of the bits that talk about what other functions are doing (e.g. `should_demote_..` in preprocessing), and made the language less assertive about what it does, since the presence of a project in this list does not imply that it's guaranteed to be in the LPQ.

**Review by loewenheim — APPROVED** [review]
